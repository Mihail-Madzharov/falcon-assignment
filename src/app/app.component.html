<swal
  #sweetAlert
  (confirm)="onSweetAlertConfirm()"
  (cancel)="onSweetAlertCancel()"
>
</swal>
<router-outlet></router-outlet>
