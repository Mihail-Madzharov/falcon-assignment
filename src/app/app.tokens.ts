import { InjectionToken } from "@angular/core";
import { Dispatcher } from "./shared/types";

export const DispatcherToken = new InjectionToken<Dispatcher>(
  "This will provide the dispatcher function"
);
