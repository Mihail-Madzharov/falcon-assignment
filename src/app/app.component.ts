import { Component, Inject, OnInit, ViewChild } from "@angular/core";
import { SweetAlertToken } from "./store/app.token";
import { Observable } from "rxjs";
import { SwalComponent } from "@sweetalert2/ngx-sweetalert2";
import { tap } from "rxjs/operators";

import { SweetAlertModel } from "./shared/sweet-allert.model";

@Component({
  selector: "app-root",
  templateUrl: "./app.component.html",
  styleUrls: ["./app.component.scss"]
})
export class AppComponent implements OnInit {
  private sweetAlertConfirmCallback: () => void;
  private cancelCallback: () => void;

  @ViewChild("sweetAlert", { static: true })
  sweetAlert: SwalComponent;

  constructor(
    @Inject(SweetAlertToken)
    public sweetAlert$: Observable<SweetAlertModel>
  ) {}

  ngOnInit(): void {
    this.sweetAlert$
      .pipe(
        tap(value => {
          if (value) {
            this.sweetAlert.options = value.options;
            this.sweetAlertConfirmCallback = value.confirmCallback;
            this.cancelCallback = value.cancelCallback;
            this.sweetAlert.show();
          }
        })
      )
      .subscribe();
  }

  onSweetAlertConfirm() {
    this.sweetAlertConfirmCallback();
  }

  onSweetAlertCancel() {
    this.cancelCallback();
  }
}
