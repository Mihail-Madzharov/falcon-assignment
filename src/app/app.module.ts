import { BrowserModule } from "@angular/platform-browser";
import { NgModule } from "@angular/core";
import { StoreModule, Store } from "@ngrx/store";
import { EffectsModule } from "@ngrx/effects";
import { SweetAlert2Module } from "@sweetalert2/ngx-sweetalert2";

import { AppRoutingModule } from "./app-routing.module";
import { AppComponent } from "./app.component";
import { HomeModule } from "./home/home.module";
import { DispatcherToken } from "./app.tokens";

@NgModule({
  declarations: [AppComponent],
  imports: [
    BrowserModule,
    AppRoutingModule,
    HomeModule,
    StoreModule.forRoot({}),
    EffectsModule.forRoot([]),
    SweetAlert2Module.forRoot()
  ],
  providers: [
    {
      provide: DispatcherToken,
      useFactory: store => action => store.dispatch(action),
      deps: [Store]
    }
  ],
  bootstrap: [AppComponent]
})
export class AppModule {}
