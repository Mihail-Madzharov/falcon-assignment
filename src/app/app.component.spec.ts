import { TestBed, async } from "@angular/core/testing";
import { RouterTestingModule } from "@angular/router/testing";
import {
  CUSTOM_ELEMENTS_SCHEMA,
  Component,
  Output,
  EventEmitter
} from "@angular/core";
import { Subject } from "rxjs";

import { AppComponent } from "./app.component";
import { SweetAlertToken } from "./store/app.token";

@Component({
  // tslint:disable-next-line:component-selector
  selector: "swal",
  template: ""
})
export class SweetAlertMockComponent {
  @Output()
  confirm = new EventEmitter();
  @Output()
  cancel = new EventEmitter();
}

describe("AppComponent", () => {
  let fixture;
  let app: AppComponent;
  let sweetAlert: Subject<any>;
  beforeEach(async(() => {
    sweetAlert = new Subject();
    TestBed.configureTestingModule({
      imports: [RouterTestingModule],
      declarations: [AppComponent, SweetAlertMockComponent],
      providers: [{ provide: SweetAlertToken, useValue: sweetAlert }]
    }).compileComponents();
    fixture = TestBed.createComponent(AppComponent);
    app = fixture.debugElement.componentInstance;
  }));

  it("should create the app", () => {
    expect(app).toBeTruthy();
  });

  it("should match snapshot", () => {
    expect(fixture).toMatchSnapshot();
  });

  it("should call show", () => {
    // Arrange
    app.ngOnInit();
    const showSpy = jest.fn();
    app.sweetAlert = { show: showSpy } as any;
    // Act
    sweetAlert.next({});

    // Assert
    expect(showSpy).toHaveBeenCalled();
  });

  it("should call the sweetAlertConfirm callBack on onSweetAlertConfirm", () => {
    // Arrange
    app.ngOnInit();
    const confirmCallbackSpy = jest.fn();
    app.sweetAlert = {
      show: () => {}
    } as any;
    // Act
    sweetAlert.next({ confirmCallback: confirmCallbackSpy });
    app.onSweetAlertConfirm();
    // Assert
    expect(confirmCallbackSpy).toHaveBeenCalled();
  });

  it("should call the sweetAlertCancelCallback on onSweetAlertCancel", () => {
    // Arrange
    app.ngOnInit();
    const cancelCallBack = jest.fn();
    app.sweetAlert = {
      show: () => {}
    } as any;
    // Act
    sweetAlert.next({ cancelCallback: cancelCallBack });
    app.onSweetAlertCancel();
    // Assert
    expect(cancelCallBack).toHaveBeenCalled();
  });
});
