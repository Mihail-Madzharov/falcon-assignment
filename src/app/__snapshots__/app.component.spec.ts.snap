// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppComponent should match snapshot 1`] = `
<app-root
  sweetAlert={[Function SweetAlertMockComponent]}
  sweetAlert$={[Function Subject]}
>
  <swal /><router-outlet />
</app-root>
`;
